File: java/IsPalindromePermutation.java
import java.util.Hashtable;

// Given a string, write a function to check if it is a permutation of a palindrome.
class IsPalindromePermutation {
    public static void main(String args[]) {
        System.out.println(isPalindromePermutation(" tenet"));
        System.out.println(isPalindromePermutation("tn eet"));
        System.out.println(isPalindromePermutation("tnE te"));
        System.out.println(isPalindromePermutation("tneeta"));
        System.out.println(isPalindromePermutation("Tact Coa"));

        System.out.println(isPalindromePermutationBitVector(" tenet"));
        System.out.println(isPalindromePermutationBitVector("tn eet"));
        System.out.println(isPalindromePermutationBitVector("tnE te"));
        System.out.println(isPalindromePermutationBitVector("tneeta"));
        System.out.println(isPalindromePermutationBitVector("Tact Coa"));
    }

    // time complexity: O(n), space complexity: O(n)
    private static boolean isPalindromePermutation(String str) {
        Hashtable<Character, Integer> characterToCount = new Hashtable<Character, Integer>();
        for (char c : str.toLowerCase().toCharArray()) {
            if (c == ' ') {
                continue;
            } else if (characterToCount.get(c) != null) {
                characterToCount.put(c, characterToCount.get(c) + 1);
            } else {
                characterToCount.put(c, 1);
            }
        }
        boolean foundOddCountChar = false;
        for (int charCount : characterToCount.values()) {
            if (charCount % 2 != 0) {
                if (foundOddCountChar) {
                    return false;
                }
                foundOddCountChar = true;
            }
        }
        return true;
    }

    // time complexity: O(n), space complexity: O(m) being m the charset length (128 for ASCII or 144.697 for UNICODE)
    private static boolean isPalindromePermutationBitVector(String str) {
        // assuming ASCII
        int charCodeToCountBitVector[] = new int[128];
        for (char c : str.toLowerCase().toCharArray()) {
            int charCode = (int) c;
            if (c == ' ') {
                continue;
            } else {
                charCodeToCountBitVector[charCode] = charCodeToCountBitVector[charCode] + 1;
            }
        }
        boolean foundOddCountChar = false;
        for (int charCount : charCodeToCountBitVector) {
            if (charCount % 2 != 0) {
                if (foundOddCountChar) {
                    return false;
                }
                foundOddCountChar = true;
            }
        }
        return true;
    }
}

File: python/is_palindrome_permutation.py
#Given a string, write a function to check if it is a permutation of a palindrome.
def get_char_to_count_map(str):
    charToCount = {}
    for char in str:
        if char == ' ':
            continue
        elif char in charToCount:
            charToCount[char] = charToCount[char] + 1
        else:
            charToCount[char] = 1
    return charToCount

# time complexity: O(n), space complexity: O(n)
def is_palindrome_permutation(str):
    charToCount = get_char_to_count_map(str)
    hasFoundOddCount = False
    for char, count in charToCount.items():
        if count % 2 != 0:
            if hasFoundOddCount:
                return False
            hasFoundOddCount = True

    return True


print(is_palindrome_permutation("tenet"))
print(is_palindrome_permutation("teent"))
print(is_palindrome_permutation("tteen"))
print(is_palindrome_permutation("teneta"))
print(is_palindrome_permutation("te net "))
